### PR TITLE
Prevent duplicate sending of headers when overridden by the user

### LINKED ISSUE #77: [Feat] Allow modification of hardcoded headers (Accepts Range, Content-Length)
### Description

Hi there, in order to support a custom file serving scenario (from UART) that also supports `HEAD` request to be able to resume downloads, I need to override two hard coded headers:

```cpp
// ...
response->acceptsRange(true);
response->sendContentLength(false);
```

Now it would be awesome if you could implement it, so that instead of using a _fork_ of a fork of a fork, this feature would exist in this repo :)

```
String AsyncWebServerResponse::_assembleHead(uint8_t version){
  if(version){
    if(!_acceptsRange)
      addHeader(F("Accept-Ranges"), F("none"));
    if(_chunked)
      addHeader(F("Transfer-Encoding"), F("chunked"));
  }
```

I can also send a PR if you will accept this. Thank you for maintaining ESPAsyncWebServer!

### Board

ESP8266

### PR DESCRIPTION
**Pull request by DRSDavidSoft**
This PR fixes #77.

### Description
These new additions can be used to design custom file serving scenario that also supports `HEAD` request to be able to resume downloads and serve 206 partial content.

### Additional notes
The user may control whether these headers are sent using the following methods:
```
response->sendAcceptsRange(false);
response->sendContentLength(false);
```

Which means that if the user sets these to `false`, they are expected to supply their own header values via `addHeader()`.

### Examples
```cpp
server->on("/file/download", HTTP_HEAD|HTTP_GET, [&](AsyncWebServerRequest *request) {
	// ...
	if (request->method() == HTTP_HEAD)
	{
		response = request->beginResponse(200, "application/octet-stream", "");
		response->setContentLength(0);
		response->sendAcceptsRange(false);
		response->sendContentLength(false);
		response->addHeader("Accept-Ranges", "bytes");
		response->addHeader("Content-Length", String((uint32_t) myFile.getSize()));
		// ...
	}
}
```

### CODE REVIEW

**Comment by mathieucarbou** [comment]
@DRSDavidSoft : can you please try a way to only modify the addHeader method to set the boolean flags when applicable? I would prefer to not leak implementation details. Users should not have to ask the library to skip a header: if we can detect that the user wants to add these headers am then the library can skip adding them

**Review by mathieucarbou — CHANGES_REQUESTED** [review]
Please see my comment around another proposal to implement that without leaking internal implementation details

**Comment by DRSDavidSoft** [comment]
@mathieucarbou That is an excellent approach, and my initial idea on how to implement it.
I'm sad to say as the user I personally had to take a look at the internal implementation details in order to figure this out and implement my use case.

I couldn't find your comment, but my original idea consisted of doing something like this:
1. User calls addHeader
2. The headers are added to the list
3. Upon sending the response, the processor keeps a flag of the sent headers. If none that matches our are sent, we will send our own.

This works great in order to avoid duplicate headers, but we have to constantly compare the header names that will be sent using something like `strcasecmp()`.

Alternatively, the following scenario could take effect: instead of comparing the header names during the send procedure, we could compare it during the add procedure.  The same string comparison will still be required, I'm afraid 🥲

This way, the unnecessary calls will be removed, so that the following code:
```cpp
response->sendAcceptsRange(false);
response->sendContentLength(false);
response->addHeader("Accept-Ranges", "bytes");
response->addHeader("Content-Length", String((uint32_t) myFile.getSize()));
```

Will be simplified to be this:
```cpp
response->addHeader("Accept-Ranges", "bytes");
response->addHeader("Content-Length", String((uint32_t) myFile.getSize()));
```

If you are okay with this, I will go the route of modifying the `addHeader()` as you requested. Cheers! 🙏

**Comment by mathieucarbou** [comment]
> Note that the string comparison would add a bit of overhead. I'm personally not against implementing this but others might disagree.
> 
> Or, maybe I misunderstood and there is some other way to go about this.

I was thinking about similar implementation using a  case insensitive comparison too. The reason I think this is acceptable is because I don't think it would add a lot of overhead on request handlers. Also, it is possible to improve the comparison by first comparing the pointers (==), then the complete string if pointers are not equal.

**Comment by DRSDavidSoft** [comment]
> I was thinking about similar implementation using a case insensitive comparison too. The reason I think this is acceptable is because I don't think it would add a lot of overhead on request handlers.

Yes, I also agree, the benefits outweigh the overhead.

> Also, it is possible to improve the comparison by first comparing the pointers (with ==, at least on esp32), then the complete string if pointers are not equal.

Oh, that's a good idea! Although for a non case sensitive comparison, the comparison must still be done.

Should I introduce a new function for the comparison, or use strcasecmp()?

File: src/WebResponses.cpp
Comment (DRSDavidSoft) [inline src/WebResponses.cpp:265]: @mathieucarbou Would this suffice? Would appreciate your feedback!

Comment (mathieucarbou) [reply src/WebResponses.cpp:265]: You can optimise the code using == to compare before using strcasecmp : if am not mistaken `T_Accept_Ranges    == "Accept-Ranges"` should be true because the compiler puts the strings at the same address, so if user passes a hardcoded string, it should be at the same address.

File: src/WebResponses.cpp
Comment (mathieucarbou) [inline src/WebResponses.cpp:280]: _sendContentType is never updated to false ? (No header check ?)

**Review by mathieucarbou — COMMENTED** [review]
A few comments on the code. I will b able to test next week on my side when I will be back from vacation.

It would be nice also to get a review from @vortigont if possible

Thanks!

File: src/WebResponses.cpp
Comment (DRSDavidSoft) [reply src/WebResponses.cpp:285]: My bad, forgot to include it. Now fixed!

**Comment by DRSDavidSoft** [comment]
Good news!

Since the header name's type is already `String`, I opted to use the existing `String::equalsIgnoreCase()` instead of C's strcasecmp. A nice side effect of this is that their implementation already includes the comparison by pointer address, in the first line!

```cpp
  if (this == &s2) {
    return true;
  }
```

here's the specific lines in both ESP cores for reference, below:

### ESP8266's implementation
https://github.com/esp8266/Arduino/blob/0e5d358c3c15cff4b12fd89d9e605ff9fa0709a6/cores/esp8266/WString.cpp#L558C6-L572

### ESP32's implementation
https://github.com/espressif/arduino-esp32/blob/1a42b87fb5a92e25596cff10ebe563a85b847fab/cores/esp32/WString.cpp#L568-L586

> You can optimize the code using == to compare before using strcasecmp : if am not mistaken `T_Accept_Ranges    == "Accept-Ranges"` should be true because the compiler puts the strings at the same address, so if user passes a hardcoded string, it should be at the same address.

Correct. Using Arduino's WString implementation will take care of this.

I believe the new implementation for this PR is more clean I prefer it to my old one as this is nicer.

> I will b able to test next week on my side when I will be back from vacation.

Oh, sorry to bother you on your vacation, thanks for taking time to review this. Enjoy the trip!

> It would be nice also to get a review from @vortigont if possible

Hopefully they will also be able to express their feedback. The code in this PR is still untested by me 😬 I guess I have to update my project from the old `2.9.3` version of your fork to the latest `3.1.5` so that I can be able to test the PR.

Hopefully, once this is merged, I will be able to directly upgrade to `3.1.6` to benefit from this.

And I hope you also accept the current state of the PR, hopefully this is acceptable. Once again, thank for the time to review this, you can't believe how relieved and excited I am that you are maintaining this library, since @me-no-dev is still busy with other @espressif project, so this library isn't getting all the love from him anymore 🥲

Have an awesome day! 😎🤘

**Comment by mathieucarbou** [comment]
Thank you! 
Will have a deeper look and test once back home.
In the meantime, if you could please Squash all your changed into only 1 commit.. thanks!

**Comment by DRSDavidSoft** [comment]
@mathieucarbou Sure, I force-pushed a squashed commit. Hopefully it is alright now. Thanks!

**Comment by mathieucarbou** [comment]
@DRSDavidSoft : I will take over the changes and implement it another way. I will notice you for the review.

**Comment by mathieucarbou** [comment]
@DRSDavidSoft : could you please have a look at PR https://github.com/mathieucarbou/ESPAsyncWebServer/pull/82 and test it ? I think it should be close to what you need and allows the user to globally control headers override.

**Pull request closed without merging**
